ENH: Faster count_neighour in cKDTree / + weighted input data

rainwoodman opened the pull request:
Pure pair-counting is barely useful. A common form of data comes with a weight for counting, e.g. galaxies come with a weight to compensate for fiber collision (more collision=> larger weight), but there must be use cases in other fields as well.

This PR adds a parameters to cKDTree.count_neighbors to allow counting using weights per data instead of bare pair-counting.

The performance improvement on regular counting is measured with the CNeighbors benchmark suite. The 'deep' test was to simulate a very deep tree by setting leaf_size to 1. The shallow test uses leaf_size=8. For deep trees and large number of bins the improvement can be huge. 

```
· Running 4 total benchmarks (2 commits * 1 environments * 2 benchmarks)
[  0.00%] · For scipy commit hash 073cea63:
[  0.00%] ·· Building for py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six...
[  0.00%] ·· Benchmarking py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[ 25.00%] ··· Running spatial.CNeighborsDeep.time_count_neighbors_deep                                                              ok
[ 25.00%] ···· 
               ================== ========= ========= ========== ========== ==========
               --                                          Nr                         
               ------------------ ----------------------------------------------------
                  (m, n1, n2)         2         10       100        400        1000   
               ================== ========= ========= ========== ========== ==========
                (2, 1000, 1000)     9.53ms   35.38ms   63.87ms    90.81ms    133.08ms 
                (8, 1000, 1000)    52.37ms   68.70ms   111.52ms   197.85ms   358.69ms 
                (16, 1000, 1000)   57.90ms   69.03ms   117.52ms   230.17ms   446.46ms 
               ================== ========= ========= ========== ========== ==========

[ 50.00%] ··· Running spatial.CNeighborsDeep.time_count_neighbors_shallow                                                           ok
[ 50.00%] ···· 
               ================== ========= ========= ========= ========== ==========
               --                                          Nr                        
               ------------------ ---------------------------------------------------
                  (m, n1, n2)         2         10       100       400        1000   
               ================== ========= ========= ========= ========== ==========
                (2, 1000, 1000)     2.53ms    8.13ms   27.72ms   64.58ms    140.12ms 
                (8, 1000, 1000)     9.27ms   14.73ms   69.47ms   204.95ms   483.66ms 
                (16, 1000, 1000)   14.38ms   19.02ms   75.66ms   228.15ms   533.98ms 
               ================== ========= ========= ========= ========== ==========

[ 50.00%] · For scipy commit hash 1eeef280:
[ 50.00%] ·· Building for py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six...
[ 50.00%] ·· Benchmarking py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[ 75.00%] ··· Running spatial.CNeighborsDeep.time_count_neighbors_deep                                                              ok
[ 75.00%] ···· 
               ================== ========= ========== ========== ========== ==========
               --                                           Nr                         
               ------------------ -----------------------------------------------------
                  (m, n1, n2)         2         10        100        400        1000   
               ================== ========= ========== ========== ========== ==========
                (2, 1000, 1000)    15.20ms   53.33ms    122.29ms   218.28ms   381.81ms 
                (8, 1000, 1000)    81.06ms   104.77ms   299.83ms   724.96ms    1.53s   
                (16, 1000, 1000)   85.85ms   109.32ms   330.75ms   847.97ms    1.85s   
               ================== ========= ========== ========== ========== ==========

[100.00%] ··· Running spatial.CNeighborsDeep.time_count_neighbors_shallow                                                           ok
[100.00%] ···· 
               ================== ========= ========= ========= ========== ==========
               --                                          Nr                        
               ------------------ ---------------------------------------------------
                  (m, n1, n2)         2         10       100       400        1000   
               ================== ========= ========= ========= ========== ==========
                (2, 1000, 1000)     3.00ms    9.29ms   31.96ms   81.50ms    171.24ms 
                (8, 1000, 1000)     9.80ms   15.73ms   72.63ms   217.93ms   511.78ms 
                (16, 1000, 1000)   15.53ms   21.28ms   79.60ms   240.17ms   560.76ms 
               ================== ========= ========= ========= ========== ==========

BENCHMARK                                    BEFORE      AFTER     FACTOR
...hbors_shallow((16, 1000, 1000), 1000)   560.76ms   533.98ms   0.95224981x
...ghbors_shallow((16, 1000, 1000), 100)    79.60ms    75.66ms   0.95050881x
...ghbors_shallow((16, 1000, 1000), 400)   240.17ms   228.15ms   0.94994971x
...neighbors_shallow((8, 1000, 1000), 2)     9.80ms     9.27ms   0.94612715x
...ghbors_shallow((8, 1000, 1000), 1000)   511.78ms   483.66ms   0.94505563x
...ighbors_shallow((8, 1000, 1000), 400)   217.93ms   204.95ms   0.94044012x
...eighbors_shallow((8, 1000, 1000), 10)    15.73ms    14.73ms   0.93628868x
...eighbors_shallow((16, 1000, 1000), 2)    15.53ms    14.38ms   0.92618798x
...ighbors_shallow((16, 1000, 1000), 10)    21.28ms    19.02ms   0.89399968x
...eighbors_shallow((2, 1000, 1000), 10)     9.29ms     8.13ms   0.87586497x
...ighbors_shallow((2, 1000, 1000), 100)    31.96ms    27.72ms   0.86724635x
...neighbors_shallow((2, 1000, 1000), 2)     3.00ms     2.53ms   0.84421965x
...ghbors_shallow((2, 1000, 1000), 1000)   171.24ms   140.12ms   0.81831769x
...ighbors_shallow((2, 1000, 1000), 400)    81.50ms    64.58ms   0.79230330x
...t_neighbors_deep((16, 1000, 1000), 2)    85.85ms    57.90ms   0.67439301x
...t_neighbors_deep((2, 1000, 1000), 10)    53.33ms    35.38ms   0.66339412x
...t_neighbors_deep((8, 1000, 1000), 10)   104.77ms    68.70ms   0.65576404x
...nt_neighbors_deep((8, 1000, 1000), 2)    81.06ms    52.37ms   0.64604150x
..._neighbors_deep((16, 1000, 1000), 10)   109.32ms    69.03ms   0.63143677x
...nt_neighbors_deep((2, 1000, 1000), 2)    15.20ms     9.53ms   0.62687735x
..._neighbors_deep((2, 1000, 1000), 100)   122.29ms    63.87ms   0.52225190x
..._neighbors_deep((2, 1000, 1000), 400)   218.28ms    90.81ms   0.41603387x
..._neighbors_deep((8, 1000, 1000), 100)   299.83ms   111.52ms   0.37193161x
...neighbors_deep((16, 1000, 1000), 100)   330.75ms   117.52ms   0.35531857x
...neighbors_deep((2, 1000, 1000), 1000)   381.81ms   133.08ms   0.34853469x
..._neighbors_deep((8, 1000, 1000), 400)   724.96ms   197.85ms   0.27290822x
...neighbors_deep((16, 1000, 1000), 400)   847.97ms   230.17ms   0.27143081x
...eighbors_deep((16, 1000, 1000), 1000)      1.85s   446.46ms   0.24083336x
...neighbors_deep((8, 1000, 1000), 1000)      1.53s   358.69ms   0.23479183x

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

rainwoodman commented:
I merged in a few performance enhancements into this PR as well. 
The weighted/unweighted counting is added via templates.

The main idea is to calculate the total weight of each tree node before the counting starts, and replace the terms in the product in traverse with the values from the weight array. The templates are used to provide support for integer (counting) and float (weighted) results.

The speed up is due to the removal of the idx array and using of binary search. The binary search insists the real_r input to the c-api is sorted. Thus we need to sort r before passing it to the c-api, then reconstruct the result. 

Insisting on a sorted c-api input also eases another patch that I've been working on, which typically gives a 4x speed up by using non-cumulative bins.

rgommers commented on scipy/spatial/tests/test_kdtree.py:
```diff
@@ -1106,6 +1106,58 @@ def test_ckdtree_memuse():
     # outside cKDTree
     assert_(num_leaks < 10)
+
+def test_ckdtree_weights ():
```
remove space before ``()`` (this causes the one issue with the TravisCI build right now, not pep8)

rgommers commented on scipy/spatial/tests/test_kdtree.py:
```diff
@@ -1106,6 +1106,58 @@ def test_ckdtree_memuse():
+
+    try:
+        tree1.build_weights(weights[:-1])
+        raise AssertionError("Excpetion is not raised for short weights list")
```
typo ``Exception``

rgommers commented:
Looks interesting. I made a few trivial comments; will leave the real review to @sturlamolden

rgommers commented:
> I merged in a few performance enhancements into this PR as well. 

Are those visible in the current asv benchmarks? If so, would be good to put the relevant asv output in the PR summary. If not, maybe add a benchmark?

rgommers commented:
Not directly related to this PR, but: isn't it about time to remove ``KDTree`` Python code and make it an alias for ``cKDTree``?

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1064,17 +1099,29 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
         p : float, optional 
             1<=p<=infinity, default 2.0
             Which Minkowski p-norm to use
+        self_weights : array_like
```
optional

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1064,17 +1099,29 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
             Which Minkowski p-norm to use
+        self_weights : array_like
+            weight of each data point in self. None is unweighted.
+        other_weights : array_like
```
optional

sturlamolden commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -236,3 +229,67 @@ build_ckdtree(ckdtree *self, npy_intp start_idx, npy_intp end_idx,
+}
+        
+        
+extern "C" PyObject*
```
Only static or inline functions in a header file. Put this in a new cxx file.

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1035,15 +1050,33 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+        cdef np.ndarray[np.float64_t, ndim=1, mode="c"] node_weights;
+        cdef np.ndarray[np.float64_t, ndim=1, mode="c"] _weights
+
+        num_of_nodes = tree_buffer_size(self.tree_buffer);
```
num_of_node = self.tree_buffer.size()

sturlamolden commented on scipy/spatial/ckdtree/src/cpp_utils.h:
```diff
@@ -38,6 +38,13 @@ tree_buffer_root(std::vector<ckdtreenode> *buf)
     return &tmp[0];
 }
 
+inline npy_intp
```
don't need this

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1064,17 +1099,29 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
             1<=p<=infinity, default 2.0
             Which Minkowski p-norm to use
+        self_weights : array_like
+            weight of each data point in self. None is unweighted.
```
Default: None.

rainwoodman commented:
@rgommers If I run python runtests.py --bench-compare master, then BEFORE is master and AFTER the the branch i am currently on?

rainwoodman commented:
I added a benchmark and addressed the inline comments (hopefully didn't miss any).  

Dropping the python implementation would likely help us to simplify the test cases. (by dropping one level of inheritance.)

rainwoodman commented:
49e6f7f is ~ 20% slower that master for very few r bins but nearly 20 times faster for large number of r bins. 

```
[  0.00%] · For scipy commit hash 49e6f7f6:
[  0.00%] ·· Building for py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six....................................................................................................................................
[  0.00%] ·· Benchmarking py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[ 25.00%] ··· Running spatial.CNeighbors.time_count_neighbors_deep                                                                  ok
[ 25.00%] ···· 
               ================== ========= ========= ========= ========= =========
               --                                         Nr                       
               ------------------ -------------------------------------------------
                  (m, n1, n2)         2         10       100       400       1000  
               ================== ========= ========= ========= ========= =========
                (2, 1000, 1000)     9.78ms   36.13ms   59.54ms   73.71ms   85.26ms 
                (8, 1000, 1000)    54.25ms   68.61ms   84.54ms   93.30ms   98.91ms 
                (16, 1000, 1000)   59.75ms   70.25ms   82.33ms   89.02ms   92.64ms 
               ================== ========= ========= ========= ========= =========

[ 50.00%] ··· Running spatial.CNeighbors.time_count_neighbors_shallow                                                               ok
[ 50.00%] ···· 
               ================== ========= ========= ========= ========= =========
               --                                         Nr                       
               ------------------ -------------------------------------------------
                  (m, n1, n2)         2         10       100       400       1000  
               ================== ========= ========= ========= ========= =========
                (2, 1000, 1000)     3.03ms    8.99ms   18.41ms   24.08ms   27.80ms 
                (8, 1000, 1000)    11.30ms   14.37ms   17.39ms   19.84ms   20.99ms 
                (16, 1000, 1000)   19.14ms   21.84ms   25.64ms   27.23ms   28.16ms 
               ================== ========= ========= ========= ========= =========

[ 50.00%] · For scipy commit hash 1eeef280:
[ 50.00%] ·· Building for py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six...
[ 50.00%] ·· Benchmarking py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[ 75.00%] ··· Running spatial.CNeighbors.time_count_neighbors_deep                                                                  ok
[ 75.00%] ···· 
               ================== ========= ========== ========== ========== ==========
               --                                           Nr                         
               ------------------ -----------------------------------------------------
                  (m, n1, n2)         2         10        100        400        1000   
               ================== ========= ========== ========== ========== ==========
                (2, 1000, 1000)    15.04ms   52.78ms    128.87ms   223.25ms   383.77ms 
                (8, 1000, 1000)    81.63ms   104.62ms   301.28ms   723.93ms    1.52s   
                (16, 1000, 1000)   85.14ms   107.74ms   333.16ms   851.38ms    1.86s   
               ================== ========= ========== ========== ========== ==========

[100.00%] ··· Running spatial.CNeighbors.time_count_neighbors_shallow                                                               ok
[100.00%] ···· 
               ================== ========= ========= ========= ========== ==========
               --                                          Nr                        
               ------------------ ---------------------------------------------------
                  (m, n1, n2)         2         10       100       400        1000   
               ================== ========= ========= ========= ========== ==========
                (2, 1000, 1000)     3.00ms    9.29ms   31.89ms   79.62ms    171.44ms 
                (8, 1000, 1000)     9.80ms   15.70ms   70.77ms   217.83ms   511.44ms 
                (16, 1000, 1000)   15.53ms   21.24ms   80.36ms   239.90ms   565.45ms 
               ================== ========= ========= ========= ========== ==========

BENCHMARK                                    BEFORE      AFTER     FACTOR
...eighbors_shallow((16, 1000, 1000), 2)    15.53ms    19.14ms   1.23225715x
...neighbors_shallow((8, 1000, 1000), 2)     9.80ms    11.30ms   1.15279087x
...eighbors_shallow((8, 1000, 1000), 10)    15.70ms    14.37ms   0.91514767x
...t_neighbors_deep((16, 1000, 1000), 2)    85.14ms    59.75ms   0.70176192x
...t_neighbors_deep((2, 1000, 1000), 10)    52.78ms    36.13ms   0.68454531x
...nt_neighbors_deep((8, 1000, 1000), 2)    81.63ms    54.25ms   0.66454037x
...t_neighbors_deep((8, 1000, 1000), 10)   104.62ms    68.61ms   0.65577264x
..._neighbors_deep((16, 1000, 1000), 10)   107.74ms    70.25ms   0.65209027x
...nt_neighbors_deep((2, 1000, 1000), 2)    15.04ms     9.78ms   0.65019373x
...ighbors_shallow((2, 1000, 1000), 100)    31.89ms    18.41ms   0.57737087x
..._neighbors_deep((2, 1000, 1000), 100)   128.87ms    59.54ms   0.46205651x
..._neighbors_deep((2, 1000, 1000), 400)   223.25ms    73.71ms   0.33014493x
...ghbors_shallow((16, 1000, 1000), 100)    80.36ms    25.64ms   0.31911796x
...ighbors_shallow((2, 1000, 1000), 400)    79.62ms    24.08ms   0.30246604x
..._neighbors_deep((8, 1000, 1000), 100)   301.28ms    84.54ms   0.28059650x
...neighbors_deep((16, 1000, 1000), 100)   333.16ms    82.33ms   0.24712316x
...ighbors_shallow((8, 1000, 1000), 100)    70.77ms    17.39ms   0.24575237x
...neighbors_deep((2, 1000, 1000), 1000)   383.77ms    85.26ms   0.22217112x
...ghbors_shallow((2, 1000, 1000), 1000)   171.44ms    27.80ms   0.16216479x
..._neighbors_deep((8, 1000, 1000), 400)   723.93ms    93.30ms   0.12887324x
...ghbors_shallow((16, 1000, 1000), 400)   239.90ms    27.23ms   0.11350268x
...neighbors_deep((16, 1000, 1000), 400)   851.38ms    89.02ms   0.10456323x
...ighbors_shallow((8, 1000, 1000), 400)   217.83ms    19.84ms   0.09106645x
...neighbors_deep((8, 1000, 1000), 1000)      1.52s    98.91ms   0.06489191x
...hbors_shallow((16, 1000, 1000), 1000)   565.45ms    28.16ms   0.04980297x
...eighbors_deep((16, 1000, 1000), 1000)      1.86s    92.64ms   0.04975783x
...ghbors_shallow((8, 1000, 1000), 1000)   511.44ms    20.99ms   0.04103185x
```

rgommers commented:
> @rgommers If I run python runtests.py --bench-compare master, then BEFORE is master and AFTER the the branch i am currently on?

I see (gh-5659) that you figured it out:)

rgommers commented:
TravisCI test failure for 419bda1 is unrelated.

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1064,17 +1110,37 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+            weight of each data point in other. 
+            If None, the pair-counting is unweighted (weighted by 1) for 'other'.
+            Default: None
+        cumulative : boolean
```
``bool, optional``

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1064,17 +1110,37 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+            The number of pairs. For unweighted counts, the result is integer.
+            For weighted counts, the result is float.
+            If cumulative is False, `result[i]` contains the counts with
+            `r[i-1] or -inf < R <= r[i]`
```
use double backticks here and in the line above, then it will be typeset as code (single backticks are references)

rainwoodman replied:
There are a few more inconsistencies in the use of quotes. for variable names 'x', do I use double or single backticks?

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1035,24 +1060,45 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+                        object self_weights=None, object other_weights=None, int cumulative=True):
         """
-        count_neighbors(self, other, r, p=2.)
+        count_neighbors(self, other, r, p=2., self_weights=None, other_weights=None, cumulative=True)
```
API question: do you think ``self_weights`` or simply ``weights`` is better?

rainwoodman replied:
Here is the thoughts I went through:

- 'weights' may confuse people when 'other' and 'self' are the same tree -- does it refer to weight on both or just on 'self'?

- 'self_weights' and 'other_weights' are too many typing and just looks very patchy;

- I considered an alternative where the tree carries the weights (such that the count_neighbors signature doesn't need to be changed), but decided against it -- it is nicer being able to reuse the same tree for different set of weights;

So I don't particularly like self_weights and other_weights neither. Any suggestions?

sturlamolden replied:
Perhaps just ```sw``` and ```ow```?

rgommers replied:
I like clearly readable names, so ``self_weights`` > ``sw``. ``weights`` seems a bit better than ``self_weights`` to me though - if the description in the docstring is clear then it shouldn't be confusing. Another alternative that avoids ``self`` would be ``tree_weights``.

sturlamolden replied:
Yes, but there are two optional weight vectors. Calling one of them "weights" will be ambiguous and also leaves the issue of what to call the other.

rainwoodman replied:
What about 
```
Long form: 
  weights=(self_weights, other_weights)

Short form:

weights=self_weights expands to 
- exception ValueError if self != other
- weights=(self_weights, self_weights) if other == self
```

rgommers replied:
I like that proposal - keeps the signature quite clean and is clear.

rainwoodman replied:
Sure. I'll put that one in and do a rebase.

rainwoodman commented:
Shall we also bring up whether to start deprecating kdtree.py  in 0.18.0?

sturlamolden commented:
I guess we should open a new issue for that.

rgommers commented:
No need for an issue I think, just needs to be proposed on the mailing list. If one of you wants to do that, that'd be helpful. Otherwise I can do it (but not this week).

rainwoodman commented:
@larsmans I replaced the function in 3ca4aa4. Thanks!

sturlamolden commented:
Yeah, an STL iterator is a templated type that has to implement the operators * and ++. If the templated container is a buffer (e.g. a raw "C" array), a pointer to the buffer is an iterator STL can use. This is a bit undercommunicated in the C++ litterature. I am not even sure all textbook authors understand it. Actually, I am quite sure many of them don't 😂

larsmans commented on scipy/spatial/ckdtree/src/count_neighbors.cxx:
```diff
@@ -29,32 +94,21 @@ traverse(const ckdtree *self, const ckdtree *other,
-            idx[n_queries] = old_idx[i];
-            ++n_queries;
-        }
+    if(end - start == 0) {
```
`if (end == start)` ?

rgommers commented:
> Shall we also bring up whether to start deprecating kdtree.py in 0.18.0?

made a proposal on scipy-dev

rgommers commented:
This PR needs a rebase. Besides the ``weights`` naming thing, is this ready to merge?

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -478,6 +498,10 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
         np.ndarray               boxsize_data
         np.float64_t             *raw_boxsize_data
 
+    property size:
```
Make this a readonly attribute of type ```np.intp_t``` instead, otherwise we must document it as a property, which makes no sense. It also requires that we change the C++ ```struct ckdtree``` accordingly. Set the value in ```__init__```.

sturlamolden replied:
We should probably also expose a count of leafnodes as a public attribute, which we get by counting them in ```_post_init```. We also need to fix the pickling to include the new attributes. If we set both their values in ```_post_init``` the pickling would work without including them. We could actually simplify the pickling of other attributes this way too.

rainwoodman replied:
It looks like a good idea. So my last commits e7beec6 is actually buggy when the tree is unpickled. 

Could you propose a good name for count of leafnodes? Why is the count of leafnode special?

sturlamolden commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -539,6 +560,9 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
             PyMem_Free(tmp)
+
+        # set the size attribute after tree_buffer is built
+        self.size = self.tree_buffer.size()
```
this must be in the function _post_init

sturlamolden replied:
because of pickling

rainwoodman commented:
I am noticing a lot of valgrind errors in varoius places. Also running vagrind the leak test fails. Some are probably not due to this PR. But I think we shall fix/understand them before merging this.

rainwoodman commented:
I still see errors with valgrind:

The first is related to pickling. There seem to be some unintialized memory used in pickling. I suspect it is the start_idx and end_idx of non-leaf nodes. Shall we set them to correct numbers as well?

The second is a failure with test_ckdtree_memuse, which fails when ran with valgrind. 

Any comments @sturlamolden ?

rainwoodman commented:
Alright, I believe we are now valgrind clean.

rainwoodman commented:
To my knowledge this PR is ready to be merged. Here is a summary of changes.

- Added weights=(a, b) parameter for weighted Pair counting

- Added 'noncumulative' for 'histogram'ing the number of pairs by distance.

- Better scaling with number of bins in pair-counting by internally always using the noncumulative bin counting. 

- Fixed a few memory access violations due to prefetching introduced in the 0.17.0 rewrite of cKDTree. [Note that as far as I know there is no benchmark backing the use of prefetch; future work is needed on this.]

- Minor tweaks to the cKDTreeNode structure: start_idx, end_idx and chilren are always properly defined even for non-leaf nodes. [Note that children, start_idx and end_idx are not independent. We may save some bytes getting rid of children]

- Added attribute .size to cKDTree, which stands for the total number of nodes in a tree. 

If there is anything we shall settle before the merge:

- A better name of cKDTree.size. The name is so suggestive that it potentially conflicts with the length of input data.
- Exposing the number of non-leaf-nodes. [It is unclear why the number of non-leaf-nodes is useful to me, and there is no clear good name for it neighter].

pv commented:
Needs rebase.
@sturlamolden: further comments?

rainwoodman commented:
I will do a rebase soon. 

But recently I am becoming a bit wary about the performance improvement of switching to non-cummulative bins. While it clearly wins when the number of bins is large, when the number of bins are small (e.g. 2~10) this is slower than the original algorithm with cumulative bins (and in some cases up to 2x or more) on practical galaxy clustering datasets.

Is it worthwhile to revive the old cumulative algorithm? A possibly reasonable approach is that the code takes the old cumulative branch is cumulative==True, and the new non-cumulative branch if cumulative==False. 

Opinions?

rgommers commented:
> > Shall we also bring up whether to start deprecating kdtree.py in 0.18.0?

> made a proposal on scipy-dev

This was okay'ed: http://thread.gmane.org/gmane.comp.python.scientific.devel/20279

rgommers commented:
> Is it worthwhile to revive the old cumulative algorithm? A possibly reasonable approach is that the code takes the old cumulative branch is cumulative==True, and the new non-cumulative branch if cumulative==False. 

More than a factor of 2x speed difference may be enough of a justification for that. That said, I'm unclear on the details and complexity of the algorithm. If you and @sturlamolden think this is a good idea, then it's fine with me.

rgommers commented:
Note that the 0.18.x branching is planned for the end of this weekend. I have time to help get this merged, but going to remove the 0.18.0 milestone now because it's not a blocker.

rainwoodman commented:
Sure. I don't think we can make it this weekend; but I'll make sure the rebase is done later of day.

The pair counting code used to be capable of doing both (cumulative and non-cumulative) with some if conditions; I think I'd put another tree walking code for the non-cumulative counting. But the verbose API between Cython and Python is slightly getting cumbersome to maintain due to

1. we insist on passing many many element types cross the interface,
2. because we have to sync struct / class types manually.

Any possibility we can improve on this regard?

rgommers commented:
Seems this is more or less done but we lost some momentum. Marked it for the next release.

The question on changing the interface is for @sturlamolden.

rgommers commented:
oh and, needs another rebase. hopefully that's the last one....

sturlamolden commented:
It is just adding two keyword arguments to the interface. Not using them makes it behave as before. the interface change seems benign. 

A minor detail though: I would prefer that the internal function ```build_weights``` not be a part of the exported Python interface. I.e. make it a ```cdef int``` with ```except -1``` instead of a ```def``` function (and remember to make it return 0 on success). Also consider to underscore the name.

rainwoodman commented:
That's a fair point. The main show stopper is that the new non-cumulative algorithm makes the code slower in certain cases and I do not have a good feeling when this happens or not (thus I don't know when to flip back to the old cumulative algorithm. Plus the cumulative algorithm was removed in a commit. 

One compromise is to say if cumulative=True then we use the cumulative algorithm, if cumulative=False then we use the non-cumulative algorithm?
The default the cumulative=True. 

I will try to settle this over the week and next.

rainwoodman commented:
@sturlamolden I renamed  `build_weights` to `_build_weights` and added documentation to claim it is internal. mostly because I prefer to have a test case for it. (it triggers a tree traversal)

rainwoodman commented:
I will add a few references on cases (mostly from cosmology) where paircounting and weighted paircounting is useful.

rainwoodman commented:
Aside from possible typos I think this is ready. Could someone remind me how to check the doc? e.g. @ev-br  or @rgommers ? Thanks.

rgommers commented:
Cythonizing ``ckdtree.pyx`` failed several times, doesn't look like a CI glitch.

For checking the docs, you can either browse to the "artifacts" tab on CircleCI, or locally build with ``make html`` in ``doc/`` and inspect the output there.

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1076,46 +1117,173 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+
+             \\xi(r) = \\frac{<D, D> - 2 f <D, R> + f^2<R, R>}{f^2<R, R>},
+
+        where the brakets represents counting pairs between two data sets
```
typo "brakets"

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1076,46 +1117,173 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+
+        Notes
+        -----
+
```
style nitpick: no blank line needed here.

rgommers commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -1076,46 +1117,173 @@ cdef public class cKDTree [object ckdtree, type ckdtree_type]:
+        ratio between number of objects from data and random.
+
+        The algorithm implemented here is loosely based on the dual-tree algorithm
+        described in [1]_. We
```
re-flow lines here

rainwoodman replied:
Hmm. How do you reflow?

I have been manually breaking lines at concept groups to keep the line short. (at least attempting to)
Is the the goal to keep the lines less than 80 char wide?

rgommers replied:
Yes indeed <80 chars. I didn't realize you broke some on purpose. This one was clearly half a sentence. I usually use ``gqj`` in vim.

rgommers replied:
changes in the last commit look good, thanks.

rgommers commented on scipy/spatial/tests/test_kdtree.py:
```diff
@@ -1139,6 +1143,80 @@ def test_ckdtree_memuse():
+        tree1._build_weights(weights[:-1])
+        raise AssertionError("Exception is not raised for short weights list")
+    except ValueError:
+        pass
```
this block can be replaced by ``assert_raises(ValueError, tree1._build_weights, weights[:-1])``

rgommers commented on scipy/spatial/tests/test_kdtree.py:
```diff
@@ -1139,6 +1143,80 @@ def test_ckdtree_memuse():
+                    weights=w1)
+            raise AssertionError("Exception is not caught")
+        except ValueError:
+            pass 
```
same here, use ``assert_raises``

rgommers commented on scipy/spatial/tests/test_kdtree.py:
```diff
@@ -1139,6 +1143,80 @@ def test_ckdtree_memuse():
+            pass 
+
+def test_ckdtree_count_neighbous_multiple_r():
+    import itertools
```
nitpick: imports at top of file

rgommers commented:
Docstring looks fine to me: https://1485-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.spatial.cKDTree.count_neighbors.html#scipy.spatial.cKDTree.count_neighbors

Last 4 commits all look good too. Made some comments but those are all minor - once addressed I think it's green-button time.

rgommers commented:
Allright, in it goes. Thanks @rainwoodman, all !
